### PR TITLE
[SPARK-41365][UI] Stages UI page fails to load for proxy in specific yarn environment

### PR DESCRIPTION
**Pull request by yabola**
### What changes were proposed in this pull request?
Stages UI page fails to load for proxy in some specific yarn environment.

### Why are the changes needed?
My environment CDH 5.8 , click to enter the spark UI from the yarn Resource Manager page
when visit the stage URI, it fails to load,  URI is
http://<yarn-url>:8088/proxy/application_1669877165233_0021/stages/stage/?id=0&attempt=0 

The issue is similar to, the final phenomenon of the issue is the same, because the parameter encode twice
[SPARK-32467](https://issues.apache.org/jira/browse/SPARK-32467)
[SPARK-33611](https://issues.apache.org/jira/browse/SPARK-33611)

The two issues solve two scenarios to avoid encode twice:
1. https redirect proxy
2. set reverse proxy enabled (spark.ui.reverseProxy)  in Nginx 

But if encode twice due to other reasons, such as this issue (yarn proxy), it will also fail when visit stage page.
It is better to decode parameter twice here. 
Just like fix here [SPARK-12708](https://issues.apache.org/jira/browse/SPARK-12708) [codes](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ui/UIUtils.scala#L626)

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
new added UT

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/status/api/v1/StagesResource.scala [hunk 1]
```diff
@@ -143,7 +148,10 @@ private[v1] class StagesResource extends BaseAppResource {
     @Context uriInfo: UriInfo):
   HashMap[String, Object] = {
     withUI { ui =>
-      val uriQueryParameters = uriInfo.getQueryParameters(true)
+      // Decode URI twice here to avoid percent-encoding twice on the query string
+      val decodeURI = URLDecoder.decode(uriInfo.getRequestUri.getRawQuery, UTF_8.name())
+      val uriQueryParameters = new ImmutableMultivaluedMap[String, String](
+        UriComponent.decodeQuery(decodeURI, true))
```
Comment (yabola) [inline core/src/main/scala/org/apache/spark/status/api/v1/StagesResource.scala:154]: Not sure if decoding twice is appropriate here.
If do not change here, the stage page can be opened, but the task sort column button cannot be clicked (locality level etc.)

File: core/src/main/scala/org/apache/spark/status/api/v1/StagesResource.scala [hunk 2]
```diff
@@ -204,7 +212,7 @@ private[v1] class StagesResource extends BaseAppResource {
       pageLength = queryParameters.getFirst("length").toInt
     }
     withUI(_.store.taskList(stageId, stageAttemptId, pageStartIndex, pageLength,
-      indexName(columnNameToSort), isAscendingStr.equalsIgnoreCase("asc")))
+      indexName(columnNameToSort), "asc".equalsIgnoreCase(isAscendingStr)))
```
Comment (yabola) [inline core/src/main/scala/org/apache/spark/status/api/v1/StagesResource.scala:215]: `isAscendingStr` may be NPE if URI was encoded twice.
After change here, at least the stage page can be opened now

**Review by yabola — COMMENTED** [review]
The change in the second place is necessary, at least the page can be opened normally and code logic has not changed.

**Comment by yabola** [comment]
Stages UI page failed 
MXADK6BDM0

**Comment by mridulm** [comment]
+CC @gengliangwang

**Comment by gengliangwang** [comment]
@yabola could you add reproduce steps or tests for the issue?

**Comment by yabola** [comment]
@gengliangwang 
My hadoop environment is CDH 5.8 ( yarn 2.6.0+cdh5.8.0+1589)
Run the latest spark code using yarn-client mode.
Click ApplicationMaster URL from ResourceManager Web UI (http://:8088/proxy/application_1669877165233_0021/stages/stage/?id=0&attempt=0)
It can reproduce easily in any application.

**Comment by yabola** [comment]
@gengliangwang 
I see comments from other people in this issues (the last one), I think he had the same problem, it is after this issues fixed time.
https://issues.apache.org/jira/browse/SPARK-32467

**Comment by yabola** [comment]
@gengliangwang Yes, but the URI will be processed by yarn proxy (encoded twice). I collect URIInfo information in the interface.

`uriInfo.getRequestUri` : `http://10.1.3.16:4040/api/v1/applications/application_1669877165233_0040/stages/0/0/taskTable?draw=2&order%255B0%255D%255Bcolumn%255D=4&order%255B0%255D%255Bdir%255D=asc&start=0&length=20&search%255Bvalue%255D=&search%255Bregex%255D=false&numTasks=21&columnIndexToSort=4&columnNameToSort=Locality%2520Level&_=1670382363732`

WEB UI request: `http://10.1.3.16:8088/proxy/application_1669877165233_0040/api/v1/applications/application_1669877165233_0040/stages/0/0/taskTable?draw=2&order%5B0%5D%5Bcolumn%5D=4&order%5B0%5D%5Bdir%5D=asc&start=0&length=20&search%5Bvalue%5D=&search%5Bregex%5D=false&numTasks=21&columnIndexToSort=4&columnNameToSort=Locality%20Level&_=1670382363732`

**Comment by gengliangwang** [comment]
@yabola I am not quite sure about the "yarn proxy" you mentioned.
Can we fix the issue in a narrow waist method? IIUC there are also ajax requests in the executor page.

**Comment by yabola** [comment]
@gengliangwang I add new UT and change decode to carefully decode each parameter, I think it aligns with the previous behavior and is more accurate ( I reuse [decodeURLParameter](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ui/UIUtils.scala#L629) , It support decode not only twice)

**Comment by yabola** [comment]
@gengliangwang If you have time, please have a look, thanks. And do I need to change codes to decode parameters only twice here?

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
@yabola Thanks for the fix. The changes LGTM.
Merging to master and branch-3.3

**Comment by gengliangwang** [comment]
@yabola There is merge conflict on branch-3.3. Could you create a backport PR against branch-3.3?

**Comment by yabola** [comment]
@gengliangwang @mridulm Thanks you all~
I created backport PR https://github.com/apache/spark/pull/39087

**Pull request closed without merging**